Instrument BundleCLI with autocomplete

kashizui opened the pull request:
Fixes #167 
Involves a major refactor of the argparse configuration. Apologies in advance, since this was necessary to build an instance of `ArgumentParser` well-formed enough to do autocompletion inference.

To test it out:
1.  Checkout the `autocomplete` branch
2.  Rerun `setup.sh`
3.  Make sure you are sourcing `rc` in your `~/bashrc`
4.  If you use `zsh`, read the comments in `rc`.

Currently only worksheet specs are being autocompleted (in addition to subcommands and flags), but please take a look at the new code structure etc. and test it out when you have time while I'm still implementing the other completers, since this is a rather hefty diff. @percyliang @pbhatnagar3

kashizui committed:
Move pip requirements to requirements.txt

kashizui committed:
Merge branch 'master' into autocomplete

kashizui committed:
Prototype argcomplete integration

kashizui committed:
Activate autocomplete in rc

kashizui committed:
Refactor bundle_cli to use subparsers and decorators

kashizui committed:
Fix help and aliases

kashizui committed:
Clean up bundle cli comments

kashizui committed:
Create BundleCLI.complete_command

kashizui committed:
Treat uuid prefixes specially in WorksheetsCompleter

kashizui committed:
Instrument cli with worksheetscompleter
and remove unnecessary nargs='?' for non-positional arguments

kashizui committed:
Fix overzealous deletion of nargs='?'

kashizui committed:
Make test-cli pass

kashizui committed:
Finish filling in worksheetscompleter

percyliang commented on rc:
```diff
@@ -13,3 +13,16 @@ clwdiff() {
+#   autoload -U +X compinit && compinit
+#   autoload -U +X bashcompinit && bashcompinit
+
+if [ -z ${CODALAB_PATH+x} ]; then
```
Let's use `dirname $0` to figure out the CODALAB_PATH so we don't have to have users set it manually.

percyliang commented:
This is great!  Structure looks good to me.

One problem: when I type 'cl work <tab>', it freezes...

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -47,60 +50,158 @@
+UUID_POST_FUNC = '[0:8]'  # Only keep first 8 characters
+
+class Commands(object):
+    '''
```
Reading this is a bit abstract, can you give an example of how Argument and Command should be used?

kashizui replied:
I'll put some comments in the code, but `Command` is used internally to represent a single CodaLab subcommand and is created by the `@command` decorator, while `Argument` is just  a temporary container for the arguments we eventually want to pass into `ArgumentParser.add_argument`.

percyliang commented on setup.sh:
```diff
@@ -21,7 +22,11 @@ if [ ! -e $env ]; then
+
+# Use sckoo@stanford.edu's patched argcomplete, until pull request has been accepted
+# https://github.com/kislyuk/argcomplete/pull/118
+$env/bin/pip install git+https://github.com/kashizui/argcomplete.git@allow-overridden-subparsersaction
```
This point is not specific to this CL, but I noticed it just now.  When I ran ./setup.sh in the codalab-cli, the webserver didn't work, because it was using a different venv.  I guess we have an implicit assumption that the codalab-cli and codalab repos share the same venv?

kashizui replied:
Ohh... that's interesting.
By the way could you explain or point me to documentation about how `codalab` is running code from my `codalab-cli` repo?

percyliang replied:
In codalab/codalab/settings/local.py, BUNDLE_SERVICE_CODE_PATH points to
the codalab-cli.

On Wed, Oct 28, 2015 at 12:53 AM, Stephen Koo notifications@github.com
wrote:

> In setup.sh
> https://github.com/codalab/codalab-cli/pull/188#discussion_r43225303:
> 
> > @@ -21,7 +22,11 @@ if [ ! -e $env ]; then
> >  fi
> > 
> >  echo "=== Install Python packages into $env..."
> > -$env/bin/pip install sqlalchemy alembic pyyaml watchdog || exit 1
> > +$env/bin/pip install -r $codalabdir/requirements.txt || exit 1
> > +
> > +# Use sckoo@stanford.edu's patched argcomplete, until pull request has been accepted
> > +# https://github.com/kislyuk/argcomplete/pull/118
> > +$env/bin/pip install git+https://github.com/kashizui/argcomplete.git@allow-overridden-subparsersaction
> 
> Ohh... that's interesting.
> By the way could you explain or point me to documentation about how
> codalab is running code from my codalab-cli repo?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/188/files#r43225303.

kashizui replied:
Ah so we extend our PYTHONPATH to include it, got it.

kashizui commented:
Hm I'm not having any trouble with `cl work`—are you on local or codalab.org?

kashizui committed:
Merge branch 'master' into autocomplete

kashizui committed:
Minor fixes for autocomplete

 - docstrings
 - use `dirname $0` in ./rc

kashizui committed:
Finish fleshing out completers

kashizui committed:
Use set -e in setup.sh

kashizui committed:
Move argcomplete requirement to requirements.txt

kashizui commented:
Finished addressing comments, and fleshing out basic implementations of all completers.
Adding `argcomplete` requirement on codalab/codalab pending:
https://github.com/codalab/codalab/pull/1361

All test-cli tests are passing, and ready for review! @percyliang @pbhatnagar3

percyliang commented:
Great!  LGTM

kashizui commented on codalab/lib/bundle_cli.py:
```diff
@@ -342,37 +435,87 @@ def get_missing_metadata(self, bundle_subclass, args, initial_metadata=None):
+        try:
+            args = parser.parse_args(argv)
+        except SystemExit as e:
+            self.exit("Try 'cl help'.")
```
check headless

kashizui commented on rc:
```diff
@@ -13,3 +13,18 @@ clwdiff() {
+codalabdir=$(dirname $0)
+
+# if [ -z ${CODALAB_PATH+x} ]; then
+#     echo "Please set your \$CODALAB_PATH to the path to your clone of the codalab-cli repository before sourcing this codalab-cli/rc file."
```
delete comment

kashizui committed:
Clean up autocomplete changes
